### PR TITLE
Added option to print test coverage and timings to st2-run-pack-tests

### PR DESCRIPTION
**Pull request by nmaludy**
This adds the ability to print test coverage and test timings to st2-run-pack-tests

Two new cmdline flags were added to support these features:

```
# -c : Run tests with code coverage reporting enabled
st2-run-pack-tests -c -p path/to/pack
```

```
# -t : Run tests with timing enabled
st2-run-pack-tests -t -p path/to/pack
```

Example output:
```
stackstorm-activedirectory/ci/virtualenv/bin/activate; \
/tmp/st2/st2common/bin/st2-run-pack-tests -c -t -x -p stackstorm-activedirectory/ci/../ || exit 1;
ENABLED COVERAGE
Running tests for pack: stackstorm-activedirectory
Installing dependencies from st2 repository...
DEPRECATION: The default format will switch to columns in the future. You can use --format=(legacy|columns) (or define a format=(legacy|columns) in your pip.conf under the [list] section) to disable this warning.                        
Installing global pack test dependencies...
Installing pack-specific dependencies...
Installing pack-specific test dependencies...
Running tests...
test_create_creds_spec (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_init (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_resolve_creds_from_cmdline (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_resolve_creds_from_cmdline_override_config (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_resolve_creds_from_config (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_resolve_creds_spec_config_creds_missing (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_resolve_creds_spec_config_creds_missing_password (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_resolve_creds_spec_config_creds_missing_username (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_resolve_creds_spec_from_cmdline (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_resolve_creds_spec_from_cmdline_override_config (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_resolve_creds_spec_from_config (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_resolve_output_ps_config (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_resolve_output_ps_config_invalid (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_resolve_output_ps_config_missing (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_resolve_output_ps_json (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_resolve_transport_cmdline (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_resolve_transport_config_base (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_resolve_transport_config_creds (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_resolve_transport_default (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_resolve_transport_kwargs_none (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_run_ad_cmdlet (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_run_ad_cmdlet_cmdlet_credentials (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_run_ad_cmdlet_cmdlet_credentials_json (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_run_ad_cmdlet_fail (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_run_ad_cmdlet_json (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_run_ad_cmdlet_json_parse_empty (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_run_ad_cmdlet_json_stderr_parse_fail (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_run_ad_cmdlet_json_stdout_parse_fail (test_action_lib_baseaction.TestActionLibBaseAction) ... passed
test_init (test_winrm_connection.TestWinRmConnection) ... passed





TEST RESULT OUTPUT:

Name                      Stmts   Miss  Cover
---------------------------------------------
lib/__init__.py               0      0   100%
lib/action.py               120     10    92%
lib/winrm_connection.py       8      2    75%
run_cmdlet.py                 6      3    50%
---------------------------------------------
TOTAL                       134     15    89%
[success] 4.29% test_action_lib_baseaction.TestActionLibBaseAction.test_run_ad_cmdlet_json_stderr_parse_fail: 0.0093s
[success] 3.96% test_action_lib_baseaction.TestActionLibBaseAction.test_run_ad_cmdlet_json_parse_empty: 0.0085s
[success] 3.94% test_action_lib_baseaction.TestActionLibBaseAction.test_run_ad_cmdlet_json: 0.0085s
[success] 3.93% test_action_lib_baseaction.TestActionLibBaseAction.test_run_ad_cmdlet_json_stdout_parse_fail: 0.0085s
[success] 3.91% test_action_lib_baseaction.TestActionLibBaseAction.test_run_ad_cmdlet_cmdlet_credentials_json: 0.0085s
[success] 3.87% test_action_lib_baseaction.TestActionLibBaseAction.test_run_ad_cmdlet_cmdlet_credentials: 0.0084s
[success] 3.84% test_action_lib_baseaction.TestActionLibBaseAction.test_run_ad_cmdlet_fail: 0.0083s
[success] 3.84% test_action_lib_baseaction.TestActionLibBaseAction.test_run_ad_cmdlet: 0.0083s
[success] 3.53% test_action_lib_baseaction.TestActionLibBaseAction.test_resolve_creds_from_config: 0.0076s
[success] 3.52% test_action_lib_baseaction.TestActionLibBaseAction.test_resolve_creds_from_cmdline: 0.0076s
[success] 3.50% test_action_lib_baseaction.TestActionLibBaseAction.test_resolve_creds_spec_from_cmdline: 0.0076s
[success] 3.49% test_action_lib_baseaction.TestActionLibBaseAction.test_create_creds_spec: 0.0075s
[success] 3.48% test_action_lib_baseaction.TestActionLibBaseAction.test_resolve_creds_spec_config_creds_missing_username: 0.0075s
[success] 3.44% test_action_lib_baseaction.TestActionLibBaseAction.test_resolve_creds_spec_config_creds_missing_password: 0.0074s
[success] 3.39% test_action_lib_baseaction.TestActionLibBaseAction.test_resolve_output_ps_config: 0.0073s
[success] 3.37% test_action_lib_baseaction.TestActionLibBaseAction.test_resolve_creds_spec_config_creds_missing: 0.0073s
[success] 3.37% test_action_lib_baseaction.TestActionLibBaseAction.test_resolve_creds_from_cmdline_override_config: 0.0073s
[success] 3.37% test_action_lib_baseaction.TestActionLibBaseAction.test_resolve_creds_spec_from_cmdline_override_config: 0.0073s
[success] 3.36% test_action_lib_baseaction.TestActionLibBaseAction.test_resolve_output_ps_config_missing: 0.0073s
[success] 3.35% test_action_lib_baseaction.TestActionLibBaseAction.test_resolve_creds_spec_from_config: 0.0072s
[success] 3.34% test_action_lib_baseaction.TestActionLibBaseAction.test_resolve_output_ps_json: 0.0072s
[success] 3.34% test_action_lib_baseaction.TestActionLibBaseAction.test_resolve_output_ps_config_invalid: 0.0072s
[success] 3.33% test_action_lib_baseaction.TestActionLibBaseAction.test_resolve_transport_config_creds: 0.0072s
[success] 3.33% test_action_lib_baseaction.TestActionLibBaseAction.test_resolve_transport_default: 0.0072s
[success] 3.32% test_action_lib_baseaction.TestActionLibBaseAction.test_resolve_transport_cmdline: 0.0072s
[success] 3.32% test_action_lib_baseaction.TestActionLibBaseAction.test_resolve_transport_config_base: 0.0072s
[success] 3.30% test_action_lib_baseaction.TestActionLibBaseAction.test_resolve_transport_kwargs_none: 0.0071s
[success] 3.28% test_action_lib_baseaction.TestActionLibBaseAction.test_init: 0.0071s
[success] 0.67% test_winrm_connection.TestWinRmConnection.test_init: 0.0014s
-----------------------------------------------------------------------------
29 tests run in 0.821 seconds (29 tests passed)
```

### CODE REVIEW

**Comment by nmaludy** [comment]
I've added a PR to st2docs to correspond with this change: https://github.com/StackStorm/st2docs/pull/535/files

I've also added a PR to the packs ci/ makefile so that packs can now take advantage of these new capabilities: https://github.com/StackStorm-Exchange/ci/pull/46

File: CHANGELOG.rst [hunk 1]
```diff
@@ -4,6 +4,10 @@ Changelog
 in development
 --------------
 
+* Added test coverage and test timing capabilities to ``st2-run-pack-tests``.
```
Comment (bigmstone) [inline CHANGELOG.rst]: Nit-picky, but can you re-locate this to the bottom. New entries traditionally go in at the bottom of the change log.

File: st2common/bin/st2-run-pack-tests [hunk 2]
```diff
@@ -37,11 +37,13 @@ PACK_TEST_PYTHON_DEPENDENCIES_NAMES=(
     'mock'
     'unittest2'
     'nose'
+    'coverage'
 )
 PACK_TEST_PYTHON_DEPENDENCIES_VERSIONS=(
     '>=2.0,<2.1'
     '>=1.1.0,<2.0'
     '>=1.3.7'
+    ''
```
Comment (bigmstone) [inline st2common/bin/st2-run-pack-tests]: Would be best to set at least a minimum version here, even if it's the latest.

File: st2common/bin/st2-run-pack-tests [hunk 3]
```diff
@@ -65,14 +69,16 @@ STACKSTORM_VIRTUALENV_PYTHON_BINARY="${STACKSTORM_VIRTUALENV_BIN}/python"
 ####################
 
 function usage() {
-    echo "Usage: $0 [-x] [-j] [-v] -p <path to pack>" >&2
-    echo "  -x : Do not create virtualenv for test for tests, e.g. when running in existing one." >&2
-    echo "  -j : Just run tests (use previously installed dependencies" >&2
+    echo "Usage: $0  [-c] [-j] [-t] [-v] [-x] -p <path to pack>" >&2
+    echo "  -c : Run tests with code coverage reporting enabled" >&2
+    echo "  -j : Just run tests. Use previously installed dependencies" >&2
```
Comment (bigmstone) [inline st2common/bin/st2-run-pack-tests]: Hmm. The old usage didn't quite get it right, but I think we can get it a bit better. Perhaps the following:

> Just run tests. Uses previously installed dependencies and virtualenv, if any, for subsequent test runs.

File: st2common/bin/st2-run-pack-tests [hunk 4]
```diff
@@ -273,7 +285,66 @@ fi
...
+
+  # base options to enable test coverage printouts
+  NOSE_OPTS+=(--with-coverage)
+
+  # Now, by default nosetests prints out test coverage for every module found
+  # in the $PYTHONPATH... as you can imagine this is not ideal and may contain
+  # a LOT of files. To only print out the modules that we care about we need to
+  # specify them on the cdmline using --cover-package=package.name
+
+  # There are two directories that way contain python files
```
Comment (bigmstone) [inline st2common/bin/st2-run-pack-tests]: That may*?

**Comment by Kami** [comment]
What bigmstone said, so far it's looking good :+1:

File: st2common/bin/st2-run-pack-tests [hunk 5]
```diff
@@ -273,7 +285,66 @@ fi
...
+  # 2) there are sub directories that contain __init__.py files that signify
+  #    to python that they're a module (ex: actions/lib/__init.py). All files
+  #    within this sub directory will be checked for coverage if we
+  #    specify the option for the base module (ex: --cover-package=lib)
+
+  # for each of the directories in the pack that may contain python files
+  for pack_python_dir in ${PACK_PYTHON_DIRS}; do
+
+    # find all *.py files in the root of the $pack_python_dir
+    for py_absolute in $(find $pack_python_dir  -mindepth 1 -maxdepth 1 -name "*.py"); do
```
Comment (Kami) [inline st2common/bin/st2-run-pack-tests]: This seems quite a bit complicated, but I know why you have to do it.

Would be good to add some test cases for it - I know it's a bit meta and complicated, but at least we should add some example pack or similar which includes package and Python file so it covers both scenarios and then we run this script on top to at least make sure it runs and doesn't fail.

Comment (nmaludy) [inline st2common/bin/st2-run-pack-tests]: After some experimentation and reading the nosetest/coverage code i've been able to greatly simplify this logic.

Now, it simply finds all sub-directories within the pack (excluding the "tests" directory) and explicitly includes those for reporting. This successfully hides the erroneous imports, simplifies the code, and should make it more future-proof.

**Comment by nmaludy** [comment]
Looks like the coverage is printing out successfully in CircleCI. You can see an example here: https://circleci.com/gh/StackStorm/st2/3909#config/containers/0

Assuming the full build passes i'll take this out of WIP status.

**Comment by Kami** [comment]
Thanks for addressing the open issues.

Besides the two small things, PR looks good to me so I will go ahead and merge it once the tests pass :)

**Comment by Kami** [comment]
Made a small change and merged it into master - 0db5da030d568ab54958f5551fb7b737af7e8461.

Thanks again.

**Pull request closed without merging**